modules/linear_boltzmann_solvers/executors/pi_keigen_smm.cc

```diff
@@ -0,0 +1,1122 @@
+  diff_solver->options.additional_options_string = diffusion_petsc_options_;
+  log.Log() << "Done creating diffusion solver.";
+
+  // Initialize the solver
+  log.Log() << "Initializing diffusion solver.";
+  diff_solver->Initialize();
+  mpi_comm.barrier();
+  log.Log() << "Done initializing diffusion solver.";
+
+  // Assemble the system
+  log.Log() << "Assembling diffusion system.";
+  std::vector<double> tmp;
+  if (diffusion_sdm_name_ == "pwld")
+    tmp.assign(pwld.GetNumLocalDOFs(diff_uk_man), 0.0);
+  else
+    tmp.assign(pwlc_ptr_->GetNumLocalAndGhostDOFs(diff_uk_man), 0.0);
+  diff_solver->AssembleAand_b(tmp);
+  log.Log() << "Done assembling diffusion system.";
+}
+
+void
+PowerIterationKEigenSMM::Execute()
+{
+  // Start transport power iterations
+  double k_eff_ell = k_eff_;
+  double k_eff_change = 1.0;
+
+  auto phi_ell = phi_old_local_;
+
+  int nit = 0;
+  while (nit < max_iters_)
+  {
+    // Compute the transport 1/k fission source
+    SetLBSFissionSource(phi_ell, /*additive=*/false);
+    Scale(q_moments_local_, 1.0 / k_eff_);
+
+    // Solve some transport inners
+    primary_ags_solver_->Setup();
+    primary_ags_solver_->Solve();
+
+    std::vector<double> phi0;
+    TransferTransportToDiffusion(phi_new_local_, phi0);
+    auto phi0_old = phi0;
+    auto phi0_m = phi0;
+
+    // Update second-moment method data
+    ComputeClosures(psi_new_local_);
+    const auto correction = ComputeSourceCorrection();
+
+    // Start diffusion power iterations
+    double lambda = k_eff_;
+    double lambda_old = k_eff_;
+    double F0_old = lbs_solver_.ComputeFissionProduction(phi_new_local_);
+
+    for (int m = 0; m < accel_pi_max_its_; ++m)
+    {
+      // Define the nodal diffusion 1/lambda fission source.
+      // This returns a local PWLD vector which must still be integrated
+      // against the diffusion test functions.
+      auto Sf = SetNodalDiffusionFissionSource(phi0_m);
+      Scale(Sf, 1.0 / lambda);
+
+      // Solve the diffusion inners. This seems quite wasteful, so
+      // only a single iteration is taken.
+      for (int i = 0; i < 1; ++i)
+      {
+        // Define the nodal diffusion scattering source.
+        // This returns a local PWLD vector which must still be integrated
+        // against the diffusion test functions.
+        auto Ss = SetNodalDiffusionScatterSource(phi0_old);
+
+        // Assemble the normal terms of the RHS of the diffusion solver,
+        // add in the second moment correction, and solve the system
+        auto b = AssembleDiffusionRHS(Sf + Ss);
+        diffusion_solver_->AddToRHS(b + correction);
+        diffusion_solver_->Solve(phi0, /*use_initial_guess=*/true);
+
+        // Bump the new solution to old
+        phi0_old = phi0;
+      }
+
+      // Compute a new diffusion eigenvalue. This requires mapping the
+      // diffusion solution back to the transport discretization.
+      std::vector<double> phi;
+      TransferDiffusionToTransport(phi0, phi);
```

Comment:
is that operation needed even if the SMM was discretized using PWLD ?

Reply:
The logic of both PWLC->PWLD and PWLD->PWLD are within this routine. A separate vector is used for the diffusion step, so it needs to be transferred back into the flux moments vector even for the same discretization.